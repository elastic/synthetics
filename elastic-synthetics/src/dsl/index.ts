import { Journey, JourneyCallback, JourneyOptions } from './journey'
import { Step, StepCallback } from './step'

class State {
    journeys: Journey[]
    currentJourney?: Journey

    constructor() {
        this.reset();
    }

    public reset() {
        this.journeys = [];
        this.currentJourney = undefined;
    }
}

export const state = new State();

export const journey = (options: JourneyOptions | string, callback: JourneyCallback) => {
    if (typeof options === 'string') {
        options = {name: options, id: options}
    }
    const j = new Journey(options, callback);
    state.journeys.push(j)
    state.currentJourney = j;

    j.callback(); // load steps
}

export const step = (name: string, callback: StepCallback) => {
    const step = new Step(name, callback);
    state.currentJourney.steps.push(step)
}