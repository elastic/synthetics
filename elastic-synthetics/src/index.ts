require('source-map-support').install();

export * from './dsl';
export * from './runner';

import { run } from './runner' ;

if (require.main === module)  {
    run()
}
