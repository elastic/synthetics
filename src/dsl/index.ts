export * from './journey';
export * from './step';
