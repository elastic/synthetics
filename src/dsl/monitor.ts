/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { createHash } from 'crypto';
import merge from 'deepmerge';
import { bold, red } from 'kleur/colors';
import {
  ThrottlingOptions,
  Location,
  ScreenshotOptions,
  Params,
  PlaywrightOptions,
} from '../common_types';
import { indent } from '../helpers';
import { LocationsMap } from '../locations/public-locations';

export type SyntheticsLocationsType = keyof typeof LocationsMap;
export const SyntheticsLocations = Object.keys(
  LocationsMap
) as SyntheticsLocationsType[];
export const ALLOWED_SCHEDULES = [
  1, 3, 5, 10, 15, 20, 30, 60, 120, 240,
] as const;

export interface AlertConfig {
  status?: {
    enabled: boolean;
  };
  tls?: {
    enabled: boolean;
  };
}

export type MonitorConfig = {
  id?: string;
  name?: string;
  type?: string;
  tags?: string[];
  schedule?: typeof ALLOWED_SCHEDULES[number];
  enabled?: boolean;
  locations?: SyntheticsLocationsType[];
  privateLocations?: string[];
  /**
   * @deprecated This option is ignored.
   * Network throttling via chrome devtools is ignored at the moment.
   * See https://github.com/elastic/synthetics/blob/main/docs/throttling.md for more details.
   */
  throttling?: boolean | ThrottlingOptions;
  screenshot?: ScreenshotOptions;
  params?: Params;
  playwrightOptions?: PlaywrightOptions;
  alert?: AlertConfig;
};

type MonitorFilter = {
  match: string;
  tags?: string[];
};

export class Monitor {
  content?: string;
  source?: Location;
  filter: MonitorFilter;
  constructor(public config: MonitorConfig = {}) {}
  /**
   * Treat the creation time config with `monitor.use` as source of truth by
   * merging the values coming from CLI and Synthetics config file
   */
  update(globalOpts: MonitorConfig = {}) {
    this.config = merge(globalOpts, this.config, {
      arrayMerge(target, source) {
        return [...new Set(source)];
      },
    });
  }

  get type() {
    return this.config.type;
  }

  setSource(source: Location) {
    this.source = source;
  }

  /**
   * The underlying journey code of the monitor
   */
  setContent(content = '') {
    this.content = content;
  }

  /**
   * If journey files are colocated within the same file during
   * push command, when we invoke synthetics from HB we rely on
   * this filter for running that specific journey alone instead of
   * all journeys on the file
   */
  setFilter(filter: MonitorFilter) {
    this.filter = filter;
  }

  /**
   * Hash is used to identify if the monitor has changed since the last time
   * it was pushed to Kibana. Change is based on three factors:
   * - Monitor configuration
   * - Code changes
   * - File path changes
   */
  hash(): string {
    const hash = createHash('sha256');
    return hash
      .update(JSON.stringify(this.config))
      .update(this.content || '')
      .update(this.source?.file || '')
      .digest('base64');
  }

  validate() {
    const schedule = this.config.schedule;
    if (ALLOWED_SCHEDULES.includes(schedule)) {
      return;
    }
    const { config, source } = this;
    let outer = bold(
      `Invalid schedule: ${schedule}, allowed values are ${ALLOWED_SCHEDULES.join(
        ','
      )}\n`
    );
    if (source) {
      const inner = `* ${config.id} - ${source.file}:${source.line}:${source.column}\n`;
      outer += indent(inner);
    }
    throw red(outer);
  }
}
