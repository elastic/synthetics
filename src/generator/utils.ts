/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { existsSync } from 'fs';
import { join } from 'path';

export async function getPackageManager(dir: string) {
  if (existsSync(join(dir, 'yarn.lock'))) {
    return 'yarn';
  }
  return 'npm';
}

export function runCommand(pkgManager: string, command: string) {
  if (pkgManager === 'yarn') {
    return `yarn ${command}`;
  }
  return `npm run ${command}`;
}

export function replaceTemplates(input: string, values: Record<string, any>) {
  for (const key in values) {
    const finalValue = values[key];
    input = input.replace(new RegExp(`'{{` + key + `}}'`, 'g'), () => {
      if (Array.isArray(finalValue)) {
        return String(finalValue)
          .split(',')
          .filter(Boolean)
          .map(f => `'${f}'`);
      }
      if (typeof finalValue == 'number') {
        return Number(finalValue);
      }
      if (typeof finalValue == 'string') {
        return `'${finalValue}'`;
      }
      return finalValue;
    });
  }
  return input;
}

export function cloudIDToKibanaURL(id: string) {
  const ID_SEPARATOR = ':';
  const VALUE_SEPARATOR = '$';
  // Ignore values before the separator
  const index = id.lastIndexOf(ID_SEPARATOR);
  if (index >= 0) {
    id = id.substring(index + 1);
  }

  // Decode the base64 string
  const decoded = Buffer.from(id, 'base64').toString();
  const words = decoded.split(VALUE_SEPARATOR);

  // construct the kibana and ES urls from decoded value
  const [host, port = '443'] = words[0].split(ID_SEPARATOR);
  const [kibanaId, kibanaPort = port] = words[2].split(ID_SEPARATOR);
  const kibanaURL = `https://${kibanaId}.${host}:${kibanaPort}`;

  return kibanaURL;
}
