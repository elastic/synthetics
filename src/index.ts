/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { runner } from './core';
import { RunOptions } from './core/runner';
import { setLogger } from './core/logger';
import { parseArgs } from './parse_args';

/* eslint-disable @typescript-eslint/no-var-requires */
require('source-map-support').install();

export async function run(options: RunOptions) {
  const cliArgs = parseArgs();
  /**
   * Use the NODE_ENV variable to control the environment if its not explicity
   * passed from either CLI or through the API
   */
  options.environment = options.environment || process.env['NODE_ENV'];
  /**
   * set up logger with appropriate file descriptor
   * to capture all the DEBUG logs when running from heartbeat
   */
  const outfd = options.outfd ?? cliArgs.outfd;
  setLogger(outfd);

  try {
    return await runner.run({
      ...options,
      headless: options.headless ?? cliArgs.headless,
      screenshots: options.screenshots ?? cliArgs.screenshots,
      dryRun: options.dryRun ?? cliArgs.dryRun,
      journeyName: options.journeyName ?? cliArgs.journeyName,
      network: options.network ?? cliArgs.network,
      pauseOnError: options.pauseOnError ?? cliArgs.pauseOnError,
      reporter: cliArgs.json && !options.reporter ? 'json' : options.reporter,
      outfd,
    });
  } catch (e) {
    console.error('Failed to run the test', e);
    process.exit(1);
  }
}

export { beforeAll, afterAll, journey, step, before, after } from './core';
