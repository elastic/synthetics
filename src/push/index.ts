/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */
import { readFile, writeFile } from 'fs/promises';
import { prompt } from 'enquirer';
import { bold, grey } from 'kleur/colors';
import {
  getLocalMonitors,
  buildMonitorSchema,
  diffMonitors as diffMonitorHashIDs,
  MonitorSchema,
} from './monitor';
import { ALLOWED_SCHEDULES, Monitor } from '../dsl/monitor';
import {
  progress,
  liveProgress,
  write,
  error,
  warn,
  indent,
  done,
  getMonitorManagementURL,
  THROTTLING_WARNING_MSG,
} from '../helpers';
import type { PushOptions, ProjectSettings } from '../common_types';
import { findSyntheticsConfig, readConfig } from '../config';
import {
  bulkDeleteMonitors,
  bulkGetMonitors,
  bulkPutMonitors,
  createMonitorsLegacy,
  CHUNK_SIZE,
} from './kibana_api';
import {
  getChunks,
  isBulkAPISupported,
  isLightweightMonitorSupported,
  logDiff,
} from './utils';

export async function push(monitors: Monitor[], options: PushOptions) {
  const duplicates = trackDuplicates(monitors);
  if (duplicates.size > 0) {
    throw error(formatDuplicateError(duplicates));
  }
  progress(
    `Pushing monitors for '${options.id}' project in kibana '${options.space}' space`
  );

  /**
   * Legacy API for kibana which does not support bulk operations
   */
  if (!isBulkAPISupported(options.kibanaVersion)) {
    return await pushLegacy(monitors, options);
  }

  const { monitors: remote } = await bulkGetMonitors(options);

  progress(`bundling ${monitors.length} monitors`);
  const schemas = await buildMonitorSchema(monitors, true);
  const local = getLocalMonitors(schemas);

  const { newIDs, changedIDs, removedIDs, unchangedIDs } = diffMonitorHashIDs(
    local,
    remote
  );
  logDiff(newIDs, changedIDs, removedIDs, unchangedIDs);

  const updatedMonitors = new Set<string>([...changedIDs, ...newIDs]);
  if (updatedMonitors.size > 0) {
    const chunks = getChunks(schemas, CHUNK_SIZE);
    for (const chunk of chunks) {
      await liveProgress(
        bulkPutMonitors(options, chunk),
        `creating or updating ${chunk.length} monitors`
      );
    }
  }

  if (removedIDs.size > 0) {
    if (updatedMonitors.size === 0 && unchangedIDs.size === 0) {
      await promptConfirmDeleteAll(options);
    }
    const chunks = getChunks(Array.from(removedIDs), CHUNK_SIZE);
    for (const chunk of chunks) {
      await liveProgress(
        bulkDeleteMonitors(options, chunk),
        `deleting ${chunk.length} monitors`
      );
    }
  }

  done(`Pushed: ${grey(getMonitorManagementURL(options.url))}`);
}

async function promptConfirmDeleteAll(options: PushOptions) {
  write('');
  const { deleteAll } = await prompt<{ deleteAll: boolean }>({
    type: 'confirm',
    skip() {
      if (options.yes) {
        this.initial = process.env.TEST_OVERRIDE ?? true;
        return true;
      }
      return false;
    },
    name: 'deleteAll',
    message: `Pushing without any monitors will delete all monitors associated with the project.\n Do you want to continue?`,
    initial: false,
  });
  if (!deleteAll) {
    throw warn('Push command Aborted');
  }
}

function trackDuplicates(monitors: Monitor[]) {
  const monitorMap: Map<string, Monitor> = new Map();
  const duplicates: Set<Monitor> = new Set();
  for (const monitor of monitors) {
    const id = monitor.config.id;
    if (monitorMap.has(id)) {
      duplicates.add(monitorMap.get(id));
      duplicates.add(monitor);
    }
    monitorMap.set(id, monitor);
  }
  return duplicates;
}

export function formatDuplicateError(monitors: Set<Monitor>) {
  let outer = bold(`Aborted: Duplicate monitors found\n`);

  let inner = '';
  for (const monitor of monitors) {
    const { config, source } = monitor;
    inner += `* ${config.id} - ${source.file}:${source.line}:${source.column}\n`;
  }
  outer += indent(inner);
  return outer;
}

const INSTALLATION_HELP = `Run 'npx @elastic/synthetics init' to create project with default settings.`;

export async function loadSettings(configPath, ignoreMissing = false) {
  try {
    const config = await readConfig(
      process.env['NODE_ENV'] || 'development',
      configPath
    );
    // Missing config file, fake throw to capture as missing file
    if (Object.keys(config).length === 0) {
      throw '';
    }
    return config.project || ({} as ProjectSettings);
  } catch (e) {
    if (!ignoreMissing) {
      throw error(`Aborted (missing synthetics config file), Project not set up correctly.

${INSTALLATION_HELP}`);
    }
  }
}

export async function validatePush(
  opts: PushOptions,
  settings: ProjectSettings
) {
  validateSettings(opts);
  await catchIncorrectSettings(settings, opts);
}

export function validateSettings(opts: PushOptions) {
  const INVALID = 'Aborted. Invalid synthetics project settings.';

  let reason = '';
  if (!opts.id) {
    reason = `Set project id via
  - CLI '--id <id>'
  - Config file 'project.id' field`;
  } else if (!opts.locations && !opts.privateLocations) {
    reason = `Set default location for all monitors via
  - CLI '--locations <values...> or --privateLocations <values...>'
  - Config file 'monitors.locations' | 'monitors.privateLocations' field`;
  } else if (!opts.schedule) {
    reason = `Set default schedule in minutes for all monitors via
  - CLI '--schedule <mins>'
  - Config file 'monitors.schedule' field`;
  } else if (opts.schedule && !ALLOWED_SCHEDULES.includes(opts.schedule)) {
    reason = `Set default schedule(${
      opts.schedule
    }) to one of the allowed values - ${ALLOWED_SCHEDULES.join(',')}`;
  }

  if (!reason) return;

  throw error(`${INVALID}

${reason}

${INSTALLATION_HELP}`);
}

async function overrideSettings(
  configPath,
  oldValue: string,
  newValue: string
) {
  const cwd = process.cwd();
  configPath = configPath ?? (await findSyntheticsConfig(cwd, cwd));
  if (!configPath) {
    throw warn(`Unable to find synthetics config file: ${configPath}`);
  }
  const config = await readFile(configPath, 'utf-8');
  const updatedConfig = config.replace(
    `id: '${oldValue}'`,
    `id: '${newValue}'`
  );
  await writeFile(configPath, updatedConfig, 'utf-8');
}

export async function catchIncorrectSettings(
  settings: ProjectSettings,
  options: PushOptions
) {
  let override = !settings.id;
  if (settings.id && settings.id !== options.id) {
    // Add an extra line to make it easier to read the prompt
    write('');
    ({ override } = await prompt<{ override: boolean }>({
      type: 'confirm',
      name: 'override',
      skip() {
        if (options.yes) {
          this.initial = process.env.TEST_OVERRIDE ?? true;
          return true;
        }
        return false;
      },
      message: `Monitors were pushed under the '${settings.id}' project. Are you sure you want to push them under the new '${options.id}' (note that this will duplicate the monitors, the old ones being orphaned)`,
      initial: false,
    }));
    if (!override) {
      throw warn('Push command Aborted');
    }
  }
  if (override) {
    await overrideSettings(options.config, settings.id, options.id);
  }
}

export async function pushLegacy(monitors: Monitor[], options: PushOptions) {
  if (isLightweightMonitorSupported(monitors, options)) {
    throw error(
      `Aborted: Lightweight monitors are not supported in ${options.kibanaVersion}. Please upgrade to 8.5.0 or above.`
    );
  }

  let schemas: MonitorSchema[] = [];
  if (monitors.length > 0) {
    progress(`bundling ${monitors.length} monitors`);
    schemas = await buildMonitorSchema(monitors, false);
    const chunks = getChunks(schemas, 10);
    for (const chunk of chunks) {
      await liveProgress(
        createMonitorsLegacy({ schemas: chunk, keepStale: true, options }),
        `creating or updating ${chunk.length} monitors`
      );
    }
  } else {
    await promptConfirmDeleteAll(options);
  }
  await liveProgress(
    createMonitorsLegacy({ schemas, keepStale: false, options }),
    `deleting all stale monitors`
  );

  done(`Pushed: ${grey(getMonitorManagementURL(options.url))}`);
}

// prints warning if any of the monitors has throttling settings enabled during push
export function warnIfThrottled(monitors: Monitor[]) {
  const throttled = monitors.some(monitor => monitor.config.throttling != null);
  if (throttled) {
    warn(THROTTLING_WARNING_MSG);
  }
}
