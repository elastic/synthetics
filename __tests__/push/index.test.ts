/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */
process.env.NO_COLOR = '1';

import { mkdir, rm, writeFile } from 'fs/promises';
import { join } from 'path';
import { Monitor } from '../../src/dsl/monitor';
import { formatDuplicateError } from '../../src/push';
import { CLIMock } from '../utils/test-config';

describe('Push', () => {
  const PROJECT_DIR = join(__dirname, 'new-project');
  const DEFAULT_ARGS = ['--auth', 'foo'];

  async function runPush(args = DEFAULT_ARGS, env?) {
    const cli = new CLIMock()
      .args(['push', ...args])
      .run({ cwd: PROJECT_DIR, env: { ...process.env, ...env } });
    expect(await cli.exitCode).toBe(1);
    return cli.stderr();
  }

  async function fakeProjectSetup(settings, monitor) {
    await mkdir(join(PROJECT_DIR, '.synthetics'), { recursive: true });
    await writeFile(
      join(PROJECT_DIR, '.synthetics', 'project.json'),
      JSON.stringify(settings, null, 2)
    );

    await writeFile(
      join(PROJECT_DIR, 'synthetics.config.ts'),
      `export default { monitor: ${JSON.stringify(monitor)} }`
    );
  }

  beforeAll(async () => {
    await mkdir(PROJECT_DIR, { recursive: true });
  });
  afterAll(async () => {
    process.env.NO_COLOR = '';
    await rm(PROJECT_DIR, { recursive: true, force: true });
  });

  it('error when auth is ignored', async () => {
    const output = await runPush([]);
    expect(output).toContain(`required option '--auth <auth>' not specified`);
  });

  it('error when project is not setup', async () => {
    const output = await runPush();
    expect(output).toContain('Aborted. Synthetics project not set up');
  });

  it('error on empty project id', async () => {
    await fakeProjectSetup({}, {});
    const output = await runPush();
    expect(output).toMatchSnapshot();
  });

  it('error on invalid location', async () => {
    await fakeProjectSetup({ project: 'test-project' }, {});
    const output = await runPush();
    expect(output).toMatchSnapshot();
  });

  it('error on invalid schedule', async () => {
    await fakeProjectSetup(
      { project: 'test-project' },
      { locations: ['test-loc'] }
    );
    const output = await runPush();
    expect(output).toMatchSnapshot();
  });

  it('abort on push with different project id', async () => {
    await fakeProjectSetup(
      { project: 'test-project' },
      { locations: ['test-loc'], schedule: 2 }
    );
    const output = await runPush(
      [...DEFAULT_ARGS, '--project', 'new-project'],
      {
        TEST_OVERRIDE: '',
      }
    );
    expect(output).toMatchSnapshot();
  });

  it('push with different id when overriden', async () => {
    await fakeProjectSetup(
      { project: 'test-project' },
      { locations: ['test-loc'], schedule: 2 }
    );
    const output = await runPush(
      [...DEFAULT_ARGS, '--project', 'new-project'],
      {
        TEST_OVERRIDE: 'true',
      }
    );
    expect(output).toContain('No Monitors found');
  });

  it('errors on duplicate monitors', async () => {
    await fakeProjectSetup(
      { project: 'test-project' },
      { locations: ['test-loc'], schedule: 2 }
    );

    await writeFile(
      join(PROJECT_DIR, 'duplicate.journey.ts'),
      `import {journey, monitor} from '../../../src';
journey('journey 1', () => monitor.use({ id: 'duplicate id' }));
journey('journey 2', () => monitor.use({ id: 'duplicate id' }));

journey('duplicate name', () => monitor.use({ schedule: 10 }));
journey('duplicate name', () => monitor.use({ schedule: 20 }));`
    );
    const output = await runPush();
    expect(output).toContain(`Aborted: Duplicate monitors found`);
    expect(output).toContain(`duplicate id`);
    expect(output).toContain(`duplicate name`);
  });

  it('format duplicate monitors', () => {
    const duplicates = new Set([
      {
        config: { id: 'test' },
        source: { file: 'journey1.ts', line: 2, column: 5 },
      },
      {
        config: { id: 'test' },
        source: { file: 'journey2.ts', line: 10, column: 5 },
      },
    ]);
    expect(formatDuplicateError(duplicates as Set<Monitor>)).toMatchSnapshot();
  });
});
