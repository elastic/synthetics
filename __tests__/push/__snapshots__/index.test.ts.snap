// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Push abort on push with different project id 1`] = `
"âš  Push command Aborted
"
`;

exports[`Push error on empty project id 1`] = `
"Aborted. Invalid synthetics project settings.

Set project id via CLI as \`--project <id>\`

Run 'npx @elastic/synthetics init' to create project with default settings.
"
`;

exports[`Push error on invalid location 1`] = `
"Aborted. Invalid synthetics project settings.

Set default location for all monitors via
  - CLI as '--locations <values...> or --privateLocations <values...>'
  - Synthetics config file 'monitors.locations' | 'monitors.privateLocations' field

Run 'npx @elastic/synthetics init' to create project with default settings.
"
`;

exports[`Push error on invalid schedule 1`] = `
"Aborted. Invalid synthetics project settings.

Set default schedule in minutes for all monitors via
  - CLI as '--schedule <mins>'
  - Synthetics config file 'monitors.schedule' field

Run 'npx @elastic/synthetics init' to create project with default settings.
"
`;

exports[`Push format duplicate monitors 1`] = `
"Aborted: Duplicate monitors found
   * test - journey1.ts:2:5
   * test - journey2.ts:10:5
   "
`;
